lucet-runtime/lucet-runtime-internals/src/sysdeps/macos.rs

```diff
@@ -118,54 +118,55 @@ pub struct ucontext_t {
     pub uc_stack: sigaltstack,
     pub uc_link: *const ucontext_t,
     pub uc_mcsize: size_t,
-    pub uc_mcontext: *const mcontext64,
+    pub uc_mcontext: *mut mcontext64,
 }
 
 #[derive(Clone, Copy, Debug)]
-pub struct UContextPtr(*const ucontext_t);
+pub struct UContextPtr(*mut ucontext_t);
 
 impl UContextPtr {
     #[inline]
-    pub fn new(ptr: *const c_void) -> Self {
+    pub fn new(ptr: *mut c_void) -> Self {
         assert!(!ptr.is_null(), "non-null context");
-        UContextPtr(ptr as *const ucontext_t)
+        UContextPtr(ptr as *mut ucontext_t)
     }
 
     #[inline]
     pub fn get_ip(self) -> *const c_void {
-        let mcontext = &unsafe { *(*self.0).uc_mcontext };
+        let mcontext = unsafe { (*self.0).uc_mcontext.as_ref().unwrap() };
         mcontext.ss.rip as *const _
     }
+
+    #[inline]
+    pub fn set_ip(self, new_ip: *const c_void) {
+        let mcontext: &mut mcontext64 = unsafe { &mut (*self.0).uc_mcontext.as_mut().unwrap() };
+        mcontext.ss.rip = new_ip as u64;
+    }
 }
 
 #[derive(Clone, Copy)]
 #[repr(C)]
 pub struct UContext {
-    context: ucontext_t,
-    mcontext: mcontext64,
+    context: *mut ucontext_t,
 }
 
 impl UContext {
     #[inline]
-    pub fn new(ptr: *const c_void) -> Self {
-        let context = *unsafe {
-            (ptr as *const ucontext_t)
-                .as_ref()
-                .expect("non-null context")
-        };
-        let mcontext = unsafe { *context.uc_mcontext };
-        UContext { context, mcontext }
+    pub fn new(ptr: *mut c_void) -> Self {
+        let context = unsafe { (ptr as *mut ucontext_t).as_mut().expect("non-null context") };
+        UContext { context }
     }
 
     pub fn as_ptr(&mut self) -> UContextPtr {
-        self.context.uc_mcontext = &self.mcontext;
-        UContextPtr::new(&self.context as *const _ as *const _)
+        // TODO: load-bearing?
+        // self.context.uc_mcontext = &self.mcontext;
```

Comment:
I don't know why we ever saved `mcontext` independently of `context`/`ucontext_t`. @jedisct1 would you happen to know these bits?

Reply:
On Apple devices, `mcontext` is a pointer, not the actual data.

Reply:
... and it's not guaranteed that the pointer leads to something that's part of the `ucontext` structure. The structure size doesn't even account for that data unless `_XOPEN_SOURCE` is defined.
But even then, if my memory serves me correctly, MacOS didn't store the data after the pointer.

Reply:
Oh I see, because this struct copied the whole `ucontext_t`, we have to fix up the `mcontext` pointer before handing out a pointer to `self`. That makes sense, and I'll delete this comment/line since `struct UContext` is just a wrapper around the `ucontext_t*` from the signal handler now.